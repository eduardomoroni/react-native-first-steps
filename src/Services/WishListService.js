import * as RealmService from '../Realm/RealmService.js'

// TODO: Create Get WishList(Create if not exists)

const createWishList = (userId: string) => {
  const wishList = {
    id: userId,
    lastUpdate: new Date(),
    lastSync: new Date()
  }

  return RealmService.create('WishList', wishList, false)
}

const getUserWishList = (userID: string) => {
  const userWishList = RealmService.objectForPrimaryKey('WishList', userID)
  // TODO: RealmService.objectForPrimaryKey is Not Synchronous, should raise a warning
  return userWishList ? userWishList : createWishList(userID)
}

const insertCardList = (cardList: any) => {
  return RealmService.create('CardList', cardList, false)
}

const updateCardList = (cardList: any, list: any) => {
  const realmRepresentation = list.filtered('multiverseid = $0', cardList.multiverseid)[0] // Ideally use find

  if (cardList.amount <= 0) {
    return RealmService.remove(realmRepresentation)
  }

  if (list.length === 0) {
    const insertedAmount = insertCardList(cardList)

    RealmService.write(() => {
      list.push(insertedAmount)
    })
  } else {
    RealmService.write(() => {
      realmRepresentation.amount = cardList.amount
    })
  }
}

const deleteWishList = (userId: string) => {
  RealmService.deleteCollectionByKey('WishList', userId)
}

const findCardListByMultiverseid = (multiverseid: number) => {
  return RealmService.findBy('CardList', `multiverseid = ${multiverseid}`)
}

export default {
  createWishList,
  updateCardList,
  insertCardList,
  findCardListByMultiverseid,
  deleteWishList,
  getUserWishList
}
