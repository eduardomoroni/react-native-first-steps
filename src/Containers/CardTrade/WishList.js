import React, { Component } from 'react'
import { StyleSheet } from 'react-native'
import { connect } from 'react-redux'
import PropTypes from 'prop-types'
import type { Dispatch } from 'redux'
import { Metrics, Colors } from '../../Styles/Themes'
import { Tabs, Tab, Icon } from 'react-native-elements'
import ListCards from '../CardSearch/ListCards'
import { AddToModal } from '../Components'
import WishListService from '../../Services/WishListService'

// TODO: REMOVE, JUST FOR TESTS
import * as RealmService from '../../Realm/RealmService'

export class WishList extends Component {
  constructor () {
    super()
    this.state = {
      selectedTab: 'want'
    }
  }

  changeTab (selectedTab) {
    this.setState({selectedTab})
  }

  render () {
    const testCard = RealmService.objectForPrimaryKey('Card', 'd6901a23503f4953dc3f643b193a7bdb31478fc2')
    const wishList = WishListService.getUserWishList(this.props.userID)
    const { want, have } = wishList
    const { selectedTab } = this.state

    return (
      <Tabs>
        <Tab
          titleStyle={styles.tabTitle}
          selectedTitleStyle={styles.selected}
          selected={selectedTab === 'want'}
          title={selectedTab === 'want' ? 'WANT' : null}
          renderIcon={() => <Icon containerStyle={styles.icon} color={Colors.gray} name='plus-one' size={33} />}
          renderSelectedIcon={() => <Icon color={Colors.lightblue} name='plus-one' size={30} />}
          onPress={() => this.changeTab('want')}>
          {/* <ListCards cards={want} /> */}
          <AddToModal list={want} card={testCard} />
        </Tab>
        <Tab
          titleStyle={styles.tabTitle}
          selectedTitleStyle={styles.selected}
          selected={selectedTab === 'have'}
          title={selectedTab === 'have' ? 'HAVE' : null}
          renderIcon={() => <Icon containerStyle={styles.icon} color={Colors.gray} name='beenhere' size={33} />}
          renderSelectedIcon={() => <Icon color={Colors.lightblue} name='beenhere' size={30} />}
          onPress={() => this.changeTab('have')}>
          <ListCards cards={have} />
        </Tab>
      </Tabs>
    )
  }
}

WishList.propTypes = {
  foo: PropTypes.any
}

WishList.defaultProps = {
  foo: ''
}

const styles = StyleSheet.create({
  container: {
    flex: 1,
    paddingTop: Metrics.navBarHeight
  },
  selected: {
    marginTop: -1,
    marginBottom: 6
  },
  icon: {
    justifyContent: 'center',
    alignItems: 'center',
    marginTop: 12
  },
  tabTitle: {
    fontWeight: 'bold',
    fontSize: 10
  }
})

const mapStateToProps = (state) => {
  const { userID } = state.user
  return {
    userID
  }
}

const mapDispatchToProps = (dispatch: Dispatch<*>) => {
  return {
  }
}

export default connect(mapStateToProps, mapDispatchToProps)(WishList)
