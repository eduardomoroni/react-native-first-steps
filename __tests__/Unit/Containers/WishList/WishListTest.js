// @flow

import 'react-native'
import React from 'react'
import { shallow } from 'enzyme'
import { WishList } from '../../../../src/Containers/CardTrade/WishList'
import WishListService from '../../../../src/Services/WishListService'

const props = {
  userID: 'FAKE_USER_ID',
  value: 'test'
}

WishListService.getUserWishList = jest.fn()

describe('<WishList />', () => {
  it('Snapshot', () => {
    const wrapper = shallow(<WishList {...props} />)
    expect(wrapper).toMatchSnapshot()
  })
})
