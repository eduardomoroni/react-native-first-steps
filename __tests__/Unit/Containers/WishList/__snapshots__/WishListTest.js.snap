// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<WishList /> Snapshot 1`] = `
<TabNavigator>
  <TabNavigatorItem
    onPress={[Function]}
    renderIcon={[Function]}
    renderSelectedIcon={[Function]}
    selected={true}
    selectedTitleStyle={
      Object {
        "marginBottom": 6,
        "marginTop": -1,
      }
    }
    title="WANT"
    titleStyle={
      Object {
        "fontSize": 10,
        "fontWeight": "bold",
      }
    }
  >
    <Connect(ListCards)
      cards={Object {}}
    />
  </TabNavigatorItem>
  <TabNavigatorItem
    onPress={[Function]}
    renderIcon={[Function]}
    renderSelectedIcon={[Function]}
    selected={false}
    selectedTitleStyle={
      Object {
        "marginBottom": 6,
        "marginTop": -1,
      }
    }
    title={null}
    titleStyle={
      Object {
        "fontSize": 10,
        "fontWeight": "bold",
      }
    }
  >
    <Connect(ListCards)
      cards={Object {}}
    />
  </TabNavigatorItem>
</TabNavigator>
`;
